### PR TITLE
Add matrix corporal

### PR DESCRIPTION
**Pull request by D4ndellion**
###### Motivation for this change
matrix-corporal is a proxy and reconciliator for matrix-synapse letting you allow/deny API endpoints, and automatically join users to rooms/communities according to a configuration policy

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/misc/matrix-corporal.nix
Comment (aanderse) [inline nixos/modules/services/misc/matrix-corporal.nix:6]: You should add some options (like `authSharedSecretFile`) for secrets so they aren't placed in `settings` and end up in the nix store.
If you do this you can have a `preStart` script to replace the secrets with `jq` at runtime and keep your secrets safe.

Let me know if you need any details on this.

File: nixos/modules/services/misc/matrix-corporal.nix
Comment (aanderse) [inline nixos/modules/services/misc/matrix-corporal.nix:27]: This is implied so you can omit.

**Review by aanderse — CHANGES_REQUESTED** [review]
I've left some comments I hope you find useful. Note that I don't know this software so if I've said anything entirely off base please feel free to correct or ignore me. If you have any questions please don't hesitate to ask.

File: nixos/modules/services/misc/matrix-corporal.nix
Comment (Ma27) [inline nixos/modules/services/misc/matrix-corporal.nix:15]: What about some sensitive defaults?

File: pkgs/servers/matrix-corporal/default.nix
Comment (Ma27) [inline pkgs/servers/matrix-corporal/default.nix:19]: * `teams.matrix.members` is already a list, so this should be assigned directly to `maintainers`.
* I'm actually against setting the full team here. Back when I introduced it, I asked a few people about the responsibilities and some were skeptical about maintaining small community-pkgs (like `mautrix-whatsapp` or, well, this one).

File: nixos/modules/services/misc/matrix-corporal.nix
Comment (dali99) [reply nixos/modules/services/misc/matrix-corporal.nix:8]: In general I really like it when modules provide this option. Makes it easy to "roll your  own" fork, or use a package from a different nix channel without having  to resort to using overlays/overriding the package set.

Im  not sure what the general rule is for having it is

File: pkgs/servers/matrix-corporal/default.nix
Comment (dali99) [reply pkgs/servers/matrix-corporal/default.nix:19]: This is fairly widely used in corporate deployments together with ma1sd, afaik but copy that. Kind of strange to have a "matrix" group that only maintains synapse though :see_no_evil:

File: nixos/modules/services/misc/matrix-corporal.nix
Comment (dali99) [reply nixos/modules/services/misc/matrix-corporal.nix:14]: I guess that would be `lib.types.attrs` ?

File: nixos/modules/services/misc/matrix-corporal.nix
Comment (aanderse) [reply nixos/modules/services/misc/matrix-corporal.nix:8]: There is no hard rule against it, your discretion.

File: nixos/modules/services/misc/matrix-corporal.nix
Comment (dali99) [reply nixos/modules/services/misc/matrix-corporal.nix:6]: oh, hmm that is pretty tricky, especially the `AuthorizationBearerToken` stuff that _can_ be part of the config file but aren't necessarily.

I'm actually not sure how to deal with that in a sensible way

**Comment by dali99** [comment]
This wasn't really ready for review, but I still appreciate it very much! I have fixed some of the things brought up!

I will fix the commits properly once this is closer to being mergeable. For now I don't want to lose the review comments by a force push :sweat_smile:

File: nixos/modules/services/misc/matrix-corporal.nix
Comment (aanderse) [reply nixos/modules/services/misc/matrix-corporal.nix:10]: Make the option `type = with types; nullOr path;` and set `default = null;`. From here you add to the `systemd` service:
```
  systemd.services.matrix-corporal = {
    ...
    preStart = optionalString (cfg.authorizationBearerTokenFile) ''
      use jq to insert AuthorizationBearerToken: 'cat ${cfg.authorizationBearerTokenFile}' into a new copy of the file under /var/lib/matrix-corporal
    '';
  };
```

Or some variation of that.

File: nixos/modules/services/misc/matrix-corporal.nix
Comment (aanderse) [reply nixos/modules/services/misc/matrix-corporal.nix:18]: Depends what you need out of the file. That might be a good fit, or `type = with types; attrsOf (oneOf [ bool int str ]);`. Or better yet take a look at https://github.com/NixOS/nixpkgs/pull/75584.

**Comment by aanderse** [comment]
https://github.com/NixOS/nixpkgs/pull/75584 has been merged and makes creating `settings` options easier. Right from the PR description you can see how adding a generic `settings` option becomes pretty straightforward:
```nix
{ lib, config, pkgs, ... }:
let
  format = pkgs.formats.json {};
in {
  options.settings = lib.mkOption {
    type = format.type;
    default = {};
  };

  config = {
    settings.foo = 10;
    settings.bar.baz = "Hello";
    environment.etc.test.source = format.generate "config.json" config.settings;
  };
}
```

I would recommend using this approach here. If interested and you have any questions, please let me know.

**Comment by dali99** [comment]
Thanks! I've already mostly used the settings options from the RFC but I've now changed the file generator and types to match the new format goodness!

**Comment by Infinisil** [comment]
Doesn't seem ready to review just yet (current version wasn't tested), I'll wait until it's not a draft anymore for a review

**Comment by dali99** [comment]
@Infinisil Yeah, sorry I'll refrain from opening a PR before it's ready for review next time. I thought the Draft status would be enough to not bother anyone :F

@aanderse I think I need some help in figuring out the secrets stuff. I'm not entirely sure how I can use DynamicUser while also reading files with correct permissions and putting them in the right places.

With the current state of the module, /var/lib/matrix-corporal isn't created, and I'm not sure why

**Comment by aanderse** [comment]
@dali99 I pinged @Infinisil so not your fault. Feel free to open draft PRs whenever you want, it won't bother anyone.

You specified `StateDirectory` so the directory should be created once the service starts. As far as secrets management with `DynamicUser` goes there is a [new feature](https://github.com/systemd/systemd/pull/16568) being added to `systemd` to fix the existing shortcomings.

**Comment by Infinisil** [comment]
No worries about pinging even when not ready for review yet btw, I'd love to review once it's ready and am glad @aanderse pinged me :)

**Pull request closed without merging**
